flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -472,40 +486,60 @@ protected void processInput(MailboxDefaultAction.Controller controller) throws E
                         new ResumeWrapper(controller.suspendDefaultAction(timer), timer)));
     }
 
-    private void resetSynchronousSavepointId(long id, boolean succeeded) {
-        if (!succeeded && activeSyncSavepointId != null && activeSyncSavepointId == id) {
-            // allow to process further EndOfPartition events
-            activeSyncSavepointId = null;
-            operatorChain.setIgnoreEndOfInput(false);
+    protected void endData() throws Exception {
+        advanceToEndOfEventTime();
+        // finish all operators in a chain effect way
+        operatorChain.finishOperators(actionExecutor);
+
+        for (ResultPartitionWriter partitionWriter : getEnvironment().getAllWriters()) {
+            partitionWriter.notifyEndOfData();
         }
-        syncSavepointId = null;
+
+        this.endOfDataReceived = true;
     }
```

Comment:
You are calling asynchronously enqueuing `endData()` in a couple of different places. I have a feeling that if there is a race condition between executing mailbox action and (source?) operator finishing on it's own it will result in calling `endData()` twice and that this method doesn't handle it well?

I think I also do not like that we are calling `endInput()` in many places. Why don't we have a common code path for calling `endInput()` when task finishes, and let's just have a different ways how task can finish? In this regard previous behaviour in `afterInvoke()` I think was a better one.

Reply:
I don't quite get the second part of the comment.

1. We're calling` endInput` in a single place. Only in the `endData` (plus the special handling for `StreamSource` which was explicitly requested so that it can be dropped along with `StreamSourceTask` and was done that way before as well). The entire purpose of this task is to call it not when the Task finishes but when it receives the `endData`.

If I understand the first part correctly you're referring solely to the `SourceOperatorStreamTask`, am I right? All other tasks call `endData` from a single path.

Reply:
BTW, it's true we called `endInput` only in `afterInvoke`, but we were setting a flag if we should actually execute it in the same number of place (more or less).

Reply:
In the 2nd part of my comment I meant `endData()`, not only `endInput()`.

> If I understand the first part correctly you're referring solely to the SourceOperatorStreamTask, am I right? All other tasks call endData from a single path.

Mostly. `SourceStreamTask` can call `endData()` (via  `LegacySourceFunctionThread#completeProcessing`) also twice in case of exceptions/interruptions, so at least not on the happy path. 

> BTW, it's true we called endInput only in afterInvoke, but we were setting a flag if we should actually execute it in the same number of places (more or less).

Maybe, but setting a flag from a couple of places is a lesser problem vs calling a major piece of lifecycle code from a couple of places.

Can we maybe at least unify `SourceStreamOperatorTask` and network input tasks? I can see `SourceStreamTask` being different here after adding `EndOfData` event, but FLIP-27 sources should have the same lifecycle and runtime handling as network inputs.

Reply:
I tried reworking it a bit. We ever call the `endData` from the default action `StreamTask::processInput` (apart from legacy sources). I also refactored the triggering of `stop-with-savepoint --drain` after thinking a bit more on https://github.com/apache/flink/pull/16589#discussion_r678955307 

PTAL and let me know what you think.